drivers/sensor/vl53l0x/vl53l0x.c

```diff
@@ -208,34 +158,40 @@ static int vl53l0x_setup_single_shot(const struct device *dev)
 	return ret;
 }
 
-
-static int vl53l0x_init(const struct device *dev)
+static int vl53l0x_start(const struct device *dev)
 {
 	struct vl53l0x_data *drv_data = dev->data;
 	int r;
 	VL53L0X_Error ret;
 	uint16_t vl53l0x_id = 0U;
 	VL53L0X_DeviceInfo_t vl53l0x_dev_info;
 
-	LOG_DBG("enter in %s", __func__);
+	LOG_DBG("[%02X] Starting", drv_data->dt_addr);
 
+	/* Pull XSHUT high to start the sensor */
 	if (drv_data->xshut.port) {
-		r = gpio_pin_configure_dt(&drv_data->xshut,
-					  GPIO_OUTPUT | GPIO_PULL_UP);
-		if (r) {
-			LOG_ERR("[%02X] Could not configure GPIO %s %d).",
-				drv_data->dt_addr,
-				drv_data->xshut.port->name,
-				drv_data->xshut.pin);
-			return -EINVAL;
+		r = gpio_pin_set_dt(&drv_data->xshut, 1);
+		if (r < 0) {
+			LOG_ERR("[%02X] Unable to start up sensor", drv_data->dt_addr);
+			return -EIO;
 		}
-
-		gpio_pin_set_dt(&drv_data->xshut, 1);
 		k_sleep(K_MSEC(2));
 	}
 
-	drv_data->vl53l0x.i2c = drv_data->i2c;
-	drv_data->vl53l0x.I2cDevAddr = drv_data->dt_addr;
+#ifdef CONFIG_VL53L0X_RECONFIGURE_ADDRESS
+	if (drv_data->dt_addr != VL53L0X_INITIAL_ADDR) {
+		ret = VL53L0X_SetDeviceAddress(&drv_data->vl53l0x, 2 * drv_data->dt_addr);
```

Comment:
Not entirely sure where this division comes from, as it has been imported like this in https://github.com/zephyrproject-rtos/hal_st/commit/7fa2cac9647643efb3128b9d75561d69f8c2a20c

My best guess is:
* In 7bits addressing, the I2C transaction mode (read/write) is given in the lowest bit of the address, the 7 upper bits being the device address (see https://en.wikipedia.org/wiki/I%C2%B2C#7-bit_addressing)
* From various examples I found across the internet, the ST HAL seems to always store the 7bits address in the upper bits, leaving the lowest one free to be marked as read or write.
* When programming the I2C address though, the device is being given its "true" 7b address, and not the 7+1 bits.
* On the other hand, the Zephyr API always use the "true" I2C address, and will handle the left shift (`<< 1` or `* 2`) when preparing the transaction; so all the functions that communicate with the sensor, ultimately resorting to `i2c_read/write`, have been patched in here.

As a conclusion, this looks like a design difference between I2C subsystem in Zephyr  and the ST HAL. Maybe what we could do would be to remove the division in https://github.com/zephyrproject-rtos/hal_st/blob/master/sensor/vl53l0x/api/core/src/vl53l0x_api.c#L360-L361, make a new release, then update and merge this PR. This could work because the function `VL53L0X_SetDeviceAddress` is not used yet in `main`, but it could potentially break user code as some project might rely on this API (not sure if possible though).

All in all, I'm not sure if this is a better solution than simply multiplying by 2 here :shrug:

Reply:
ok, thanks for explanation. Somehow I missed that it was already there.
Let's keep it that way.